lib/dRep/dRepImpl.py

```diff
@@ -0,0 +1,441 @@
+# -*- coding: utf-8 -*-
+#BEGIN_HEADER
+import logging
+import os
+import sys
+import shutil
+import subprocess
+import pprint
+import uuid
+import re
+import functools
+
+from installed_clients.KBaseReportClient import KBaseReport
+from installed_clients.DataFileUtilClient import DataFileUtil
+from installed_clients.WorkspaceClient import Workspace
+from installed_clients.MetagenomeUtilsClient import MetagenomeUtils
+
+from .util import report
+from .util.dprint import dprint
+from .util.kbase_obj import BinnedContigs
+from .util.config import _globals, reset
+
+#END_HEADER
+
+
+class dRep:
+    '''
+    Module Name:
+    dRep
+
+    Module Description:
+    A KBase module: dRep
+    '''
+
+    ######## WARNING FOR GEVENT USERS ####### noqa
+    # Since asynchronous IO can lead to methods - even the same method -
+    # interrupting each other, you must be *very* careful when using global
+    # state. A method could easily clobber the state set by another while
+    # the latter method is running.
+    ######################################### noqa
+    VERSION = "0.0.1"
+    GIT_URL = "https://github.com/n1mus/dRep.git"
+    GIT_COMMIT_HASH = "05f421114e50a04008d9897cdb20399b23a47b46"
+
+    #BEGIN_CLASS_HEADER
+    #END_CLASS_HEADER
+
+    # config contains contents of config file in a hash or None if it couldn't
+    # be found
+    def __init__(self, config):
+        #BEGIN_CONSTRUCTOR
+
+        callback_url = os.environ['SDK_CALLBACK_URL']
+        workspace_url = config['workspace-url']
+        shared_folder = config['scratch']
+        
+        self._globals = { # shared by all API-method runs
+            'shared_folder': config['scratch'], 
+            'ws': Workspace(workspace_url),
+            'dfu': DataFileUtil(callback_url),
+            'mgu': MetagenomeUtils(callback_url),
+            'kbr': KBaseReport(callback_url),
+            }
+
+        logging.basicConfig(format='%(created)s %(levelname)s: %(message)s',
+                            level=logging.INFO)
+        #END_CONSTRUCTOR
+        pass
+
+
+    def dereplicate(self, ctx, params):
+        """
+        This example function accepts any number of parameters and returns results in a KBaseReport
+        :param params: instance of mapping from String to unspecified object
+        :returns: instance of type "ReportResults" -> structure: parameter
+           "report_name" of String, parameter "report_ref" of String
+        """
+        # ctx is the context object
+        # return variables are: output
+ 
+        #BEGIN dereplicate
+
+        dprint('params', run=locals())
+
+
+        # set up globals ds `_globals` for this API-method run
+
+        reset(_globals) # clear all fields but `debug`
```

Comment:
I would argue that you should just use self._globals and pass it down to child functions to avoid the use of any global data.

Reply:
then i'd have to pass it to other modules and that would clutter up function calls

i'm actually not sure how Impl instantiations are used outside of testing though

Reply:
my guess is your container would call my container, instantiate the Impl, and then the rest of the calls go to that instance

in testing, there's a new instance created for each run